mcos/de_noising/de_noise.py

```diff
@@ -0,0 +1,179 @@
+from sklearn.neighbors import KernelDensity
+from scipy.optimize import minimize
+import numpy as np
+import pandas as pd
+
+
+def de_noise_covariance_matrix(covariance_matrix: np.array, n_observations: int, bandwidth: float = .25) -> np.array:
+    """
+    Computes the correlation matrix associated with a given covariance matrix,
+    and derives the eigenvalues and eigenvectors for that correlation matrix.
+    Then shrinks the eigenvalues associated with noise, resulting in a de-noised correlation matrix
+    which is then used to recover the covariance matrix.
+
+    In summary, this step shrinks only the eigenvalues
+    associated with noise, leaving the eigenvalues associated with signal unchanged.
+
+    For more info see section 4.2 of "A Robust Estimator of the Efficient Frontier",
+    this function and the functions it calls are all modified from this section
+
+    :param covariance_matrix: the covariance matrix we want to de-noise
+    :param n_observations: the number of observations used to create the covariance matrix
+    :param bandwidth: bandwidth hyper-parameter for KernelDensity
+    :return: de-noised covariance matrix
+    """
+    #  q=T/N where T=sample length and N=number of variables
+    q = n_observations / covariance_matrix.shape[1]
```

Comment:
Is there a more descriptive name we could use instead of `q`?

Reply:
In the paper it's only ever referred to as `q` (with a comment saying `# q=T/N`), so I'm not sure.